src/axom/core/memory_management.hpp

```diff
@@ -200,23 +194,47 @@ inline void deallocate( T*& pointer ) noexcept
 template < typename T >
 inline T* reallocate( T* pointer, std::size_t n ) noexcept
 {
-  if ( n == 0 )
-  {
-    axom::deallocate( pointer );
-    return nullptr;
-  }
-
   const std::size_t numbytes = n * sizeof( T );
 
 #ifdef AXOM_USE_UMPIRE
 
+  // Workaround for bug in Umpire's handling on reallocate(0)
+  // Fixed in Umpire PR #292 (after v1.1.0)
+  if(n==0)
+  {
+    axom::deallocate<T>(pointer);
+    pointer = axom::allocate<T>(0);
+    return pointer;
+  }
+
   umpire::ResourceManager& rm = umpire::ResourceManager::getInstance();
+
+  // Workaround for bug in Umpire's handling of reallocate 
+  // called on a zero-sized allocation
+  // Fixed in Umpire PR #292 (after v1.1.0)
+  if(pointer != nullptr)
+  {
+    auto* allocRecord = rm.findAllocationRecord(pointer);
+    if(allocRecord && allocRecord->size == 0)
+    {
+      axom::deallocate<T>(pointer);
+      pointer = axom::allocate<T>(n);
+      return pointer;
+    }
+  }
+
   pointer = static_cast< T* >( rm.reallocate( pointer, numbytes ) );
 
 #else
 
   pointer = static_cast< T* >( std::realloc( pointer, numbytes ) );
 
+  // Consistently handle realloc(0) for std::realloc to match umpire's behavior
+  if(n==0 && pointer == nullptr)
```

Comment:
Note: This ensures that the result of ``axom::reallocate(0)`` is the same, regardless of whether Axom is configured with Umpire.